File: android/app/src/main/java/com/laneshadow/ui/components/molecules/CreativeLabelFadeIn.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * CreativeLabelFadeIn molecule component
 *
 * Fade-in animation for creative route labels with staggered reveal for multi-line labels
 * and a highlight pulse on first appearance. Smooth transition from skeleton state to final label.
 *
 * Typography: theme.type.display.md for the creative label
 * Animation: 300ms fade-in, 100ms stagger between lines, 500ms highlight pulse
 *
 * @param label The creative label text to display
 * @param visible Whether the label is ready (loaded from enrichment)
 * @param subtitle Optional subtitle or secondary line
 * @param staggerDelay Delay between line animations in ms (default 100)
 * @param fadeDuration Fade-in duration in ms (default 300)
 * @param highlightDuration Highlight pulse duration in ms (default 500)
 * @param modifier Modifier for the component
 * @param testId Test ID for UI testing
 */
@Composable
fun CreativeLabelFadeIn(
    label: String,
    visible: Boolean,
    subtitle: String? = null,
    staggerDelay: Int = 100,
    fadeDuration: Int = 300,
    highlightDuration: Int = 500,
    modifier: Modifier = Modifier,
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current
    val isInInspectionMode = LocalInspectionMode.current

    // Animation states
    var labelAlpha by remember { mutableFloatStateOf(0f) }
    var subtitleAlpha by remember { mutableFloatStateOf(0f) }
    var scale by remember { mutableFloatStateOf(1f) }
    var hasAnimatedHighlight by remember { mutableStateOf(false) }

    // Trigger animations when visible becomes true
    LaunchedEffect(visible) {
        if (visible) {
            // Animate label fade-in
            labelAlpha = 1f

            // Animate subtitle fade-in with stagger delay
            if (subtitle != null) {
                kotlinx.coroutines.delay(staggerDelay.toLong())
                subtitleAlpha = 1f
            }

            // Trigger highlight pulse animation (scale 1.0 -> 1.02 -> 1.0)
            if (!hasAnimatedHighlight) {
                hasAnimatedHighlight = true
                scale = 1.02f
                kotlinx.coroutines.delay((highlightDuration / 2).toLong())
                scale = 1f
            }
        }
    }

    // Build accessibility description
    val accessibilityDescription = "Route name: $label"

    // Skeleton placeholder when not visible
    if (!visible) {
        Column(
            modifier = modifier
                .testTag(testId ?: "creative-label-skeleton")
                .semantics {
                    contentDescription = "Loading route name"
                }
                .fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(theme.space.xs),
        ) {
            // Label skeleton
            SkeletonBar(
                width = 0.7f,
                height = 60.dp, // Approximate height for display.md text
                theme = theme,
                testTag = testId?.let { "$it-label-skeleton" } ?: "creative-label-skeleton-label",
            )

            // Subtitle skeleton (if subtitle provided)
            if (subtitle != null) {
                SkeletonBar(
                    width = 0.5f,
                    height = 24.dp, // Approximate height for body.md text
                    theme = theme,
                    testTag = testId?.let { "$it-subtitle-skeleton" } ?: "creative-label-skeleton-subtitle",
                )
            }
        }
        return
    }

    // Animated content when visible
    Column(
        modifier = modifier
            .testTag(testId ?: "creative-label")
            .semantics {
                contentDescription = accessibilityDescription
            }
            .scale(scale) // Apply highlight scale animation
            .fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(theme.space.xs),
    ) {
        // Main label line
        Text(
            text = label,
            style = theme.type.display.md,
            color = theme.colors.onSurface.default,
            modifier = Modifier
                .alpha(labelAlpha)
                .testTag(testId?.let { "$it-text" } ?: "creative-label-text"),
        )

        // Optional subtitle with stagger delay
        if (subtitle != null) {
            Text(
                text = subtitle,
                style = theme.type.body.md,
                color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                modifier = Modifier
                    .alpha(subtitleAlpha)
                    .testTag(testId?.let { "$it-subtitle" } ?: "creative-label-subtitle"),
            )
        }
    }
}

/**
 * Skeleton bar component for placeholder state
 *
 * @param width Bar width as fraction of parent
 * @param height Bar height
 * @param theme LaneShadow theme values
 * @param modifier Modifier for the bar
 * @param testTag Test ID for UI testing
 */
@Composable
private fun SkeletonBar(
    width: Float,
    height: Dp,
    theme: com.laneshadow.theme.LaneShadowThemeValues,
    modifier: Modifier = Modifier,
    testTag: String? = null,
) {
    Row(
        modifier = modifier
            .fillMaxWidth(width)
            .height(height)
            .background(
                color = theme.colors.muted.default,
                shape = RoundedCornerShape(theme.radius.sm),
            )
            .testTag(testTag ?: "skeleton-bar"),
    ) {}
}

File: ios/LaneShadow/Views/Molecules/CreativeLabelFadeIn.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Creative Label Fade In Component

/**
 * Fade-in animation for creative route labels
 *
 * Staggered reveal for multi-line labels with smooth transition from skeleton state.
 * Mimics the React Native component behavior from react-native/components/enrichment/creative-label-fade-in.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Label text: `theme.colors.onSurface.default`
 *   - Subtitle text: `theme.colors.onSurface.subtle` (with opacity)
 *   - Skeleton fill: `theme.colors.muted.default`
 * - Typography:
 *   - Label: `theme.type.display.md` (large display text)
 *   - Subtitle: `theme.type.body.md` (body text)
 *
 * ## Behavior
 * - When `visible` becomes true, fades in with opacity animation (0.3s default)
 * - If subtitle exists, staggers it by `staggerDelay` seconds (0.1s default)
 * - Shows skeleton placeholder when not visible
 * - Respects reduce-motion preference (instant reveal when enabled)
 *
 * ## Parameters
 * - label: The creative label text to display
 * - visible: Whether the label is loaded and ready to show
 * - subtitle: Optional secondary line
 * - staggerDelay: Delay between line animations in seconds (default 0.1)
 * - fadeDuration: Fade-in duration in seconds (default 0.3)
 *
 * ## Accessibility
 * - Screen reader announces "Route name: {label}" via accessibilityLabel
 */
public struct LSCreativeLabelFadeIn: View {
    @Environment(\.theme) private var theme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    @State private var labelVisible = false
    @State private var subtitleVisible = false

    private let label: String
    private let visible: Bool
    private let subtitle: String?
    private let staggerDelay: Double
    private let fadeDuration: Double

    public init(
        label: String,
        visible: Bool,
        subtitle: String? = nil,
        staggerDelay: Double = 0.1,
        fadeDuration: Double = 0.3
    ) {
        self.label = label
        self.visible = visible
        self.subtitle = subtitle
        self.staggerDelay = staggerDelay
        self.fadeDuration = fadeDuration
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: theme.space.xs) {
            if visible {
                // Main label
                Text(label)
                    .font(theme.type.display.md.font)
                    .foregroundStyle(theme.colors.onSurface.default)
                    .opacity(labelVisible ? 1 : 0)
                    .animation(.easeIn(duration: reduceMotion ? 0 : fadeDuration), value: labelVisible)
                    .accessibilityLabel("Route name: \(label)")
                    .accessibilityAddTraits(.isHeader)

                // Optional subtitle with stagger delay
                if let subtitle {
                    Text(subtitle)
                        .font(theme.type.body.md.font)
                        .foregroundStyle(
                            theme.colors.onSurface.subtle.opacity(0.7)
                        )
                        .opacity(subtitleVisible ? 1 : 0)
                        .animation(.easeIn(duration: reduceMotion ? 0 : fadeDuration), value: subtitleVisible)
                        .accessibilityLabel(subtitle)
                }
            } else {
                // Skeleton placeholder
                VStack(alignment: .leading, spacing: theme.space.xs) {
                    // Label skeleton (estimate width based on text length)
                    skeletonBar(width: estimatedLabelWidth, height: theme.type.display.md.fontSize)

                    // Subtitle skeleton
                    if subtitle != nil {
                        skeletonBar(width: estimatedSubtitleWidth, height: theme.type.body.md.fontSize)
                    }
                }
                .accessibilityLabel("Loading")
                .accessibilityElement(children: .ignore)
            }
        }
        .onAppear {
            if visible {
                showContent()
            }
        }
        .onChange(of: visible) { _, newValue in
            if newValue {
                showContent()
            } else {
                hideContent()
            }
        }
    }

    // MARK: - Private Views

    private func skeletonBar(width: CGFloat, height: CGFloat) -> some View {
        RoundedRectangle(cornerRadius: 8, style: .continuous)
            .fill(theme.colors.muted.default)
            .frame(width: width, height: height)
    }

    // MARK: - Helpers

    private func showContent() {
        // Show label immediately
        if reduceMotion {
            labelVisible = true
            subtitleVisible = true
        } else {
            labelVisible = true

            // Show subtitle after stagger delay
            if subtitle != nil {
                DispatchQueue.main.asyncAfter(deadline: .now() + staggerDelay) {
                    subtitleVisible = true
                }
            } else {
                subtitleVisible = true
            }
        }
    }

    private func hideContent() {
        labelVisible = false
        subtitleVisible = false
    }

    /// Estimate label width based on character count (rough approximation)
    private var estimatedLabelWidth: CGFloat {
        let charWidth = theme.type.display.md.fontSize * 0.6
        let minWidth: CGFloat = 120
        let maxWidth: CGFloat = 280
        let estimated = CGFloat(label.count) * charWidth
        return min(max(estimated, minWidth), maxWidth)
    }

    /// Estimate subtitle width based on character count
    private var estimatedSubtitleWidth: CGFloat {
        let charWidth = theme.type.body.md.fontSize * 0.5
        let minWidth: CGFloat = 80
        let maxWidth: CGFloat = 200
        let estimated = CGFloat(subtitle?.count ?? 0) * charWidth
        return min(max(estimated, minWidth), maxWidth)
    }
}

// MARK: - Preview

#Preview("CreativeLabelFadeIn - Visible") {
    VStack(alignment: .leading, spacing: 24) {
        LSCreativeLabelFadeIn(
            label: "Sunset Coastal Route",
            visible: true
        )

        LSCreativeLabelFadeIn(
            label: "Mountain Pass Adventure",
            visible: true,
            subtitle: "45 miles • 2 hours"
        )

        LSCreativeLabelFadeIn(
            label: "Scenic Valley Loop",
            visible: true,
            subtitle: "Intermediate terrain",
            staggerDelay: 0.2,
            fadeDuration: 0.5
        )
    }
    .padding()
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Color.gray.opacity(0.1))
    .laneShadowTheme()
}

#Preview("CreativeLabelFadeIn - Skeleton") {
    VStack(alignment: .leading, spacing: 24) {
        LSCreativeLabelFadeIn(
            label: "Sunset Coastal Route",
            visible: false
        )

        LSCreativeLabelFadeIn(
            label: "Mountain Pass Adventure",
            visible: false,
            subtitle: "45 miles • 2 hours"
        )
    }
    .padding()
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Color.gray.opacity(0.1))
    .laneShadowTheme()
}

#Preview("CreativeLabelFadeIn - Transition") {
    struct TransitionDemo: View {
        @State private var visible = false

        var body: some View {
            VStack(alignment: .leading, spacing: 24) {
                LSCreativeLabelFadeIn(
                    label: "Sunset Coastal Route",
                    visible: visible,
                    subtitle: "45 miles • 2 hours"
                )

                Button("Toggle Visibility") {
                    withAnimation(.easeInOut(duration: 0.3)) {
                        visible.toggle()
                    }
                }
                .buttonStyle(.bordered)
            }
            .padding()
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(Color.gray.opacity(0.1))
            .laneShadowTheme()
        }
    }

    return TransitionDemo()
}

#Preview("CreativeLabelFadeIn - Reduce Motion") {
    VStack(alignment: .leading, spacing: 24) {
        LSCreativeLabelFadeIn(
            label: "Sunset Coastal Route",
            visible: true,
            subtitle: "45 miles • 2 hours"
        )
    }
    .padding()
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Color.gray.opacity(0.1))
    .laneShadowTheme()
    .accessibilityReduceMotion(true)
}
